go/tools/builders/compilepkg.go

```diff
@@ -449,18 +452,25 @@ func compileArchive(
 		// Add unknown origin source files into the mix.
 		nogoSrcs = append(nogoSrcs, goSrc)
 	}
-	if nogoPath != "" && len(nogoSrcs) > 0 {
-		ctx, cancel := context.WithCancel(context.Background())
-		nogoChan = make(chan error)
-		go func() {
-			nogoChan <- runNogo(ctx, workDir, nogoPath, nogoSrcs, deps, packagePath, importcfgPath, outFactsPath)
-		}()
-		defer func() {
-			if nogoChan != nil {
-				cancel()
-				<-nogoChan
-			}
-		}()
+	if nogoPath != "" {
+		nogoLogFile, err := os.OpenFile(nogoLogPath, os.O_CREATE|os.O_RDWR, 0644)
+		if err != nil {
+			return fmt.Errorf("error opening nogo log file: %v", err)
+		}
+		defer nogoLogFile.Close()
+		if len(nogoSrcs) > 0 {
+			ctx, cancel := context.WithCancel(context.Background())
+			nogoChan = make(chan error)
+			go func() {
+				nogoChan <- runNogo(ctx, workDir, nogoPath, nogoSrcs, deps, packagePath, importcfgPath, outFactsPath, nogoLogFile)
```

Comment:
Since `runNogo` effectively just runs a separate binary, have you considered moving that invocation into the validation action? That would allow us to fully remove nogo from the critical path. We may also not need the shell post-processing anymore.

Reply:
I will look into this as a follow-up.